snuba/query/matchers.py

```diff
@@ -0,0 +1,317 @@
+        return ret
+
+    def string(self, name: str) -> str:
+        """
+        Returns a string present in the result, guaranteeing the string is there or throws.
+        """
+        ret = self.results[name]
+        assert isinstance(ret, str)
+        return ret
+
+    # TODO: Consider adding more utility method to assert and return specific types
+    # from the Results if needed.
+
+    def merge(self, values: MatchResult) -> MatchResult:
+        return MatchResult(ChainMap(self.results, values.results))
+
+
+class Pattern(ABC, Generic[TMatchedType]):
+    """
+    Tries to match a given node (like an AST Expression) with the rules provided
+    when instantiating this class. This is meant to work like a regular expression
+    but for AST expressions.
+
+    When a node is provided to this method, if this method returns an instance of
+    MatchResult, it means the node fully matches the pattern.
+    If the pattern defines named parameters, their value is returned as part of
+    the MatchResult object returned.
+    More on how to define parameters in the Param class.
+
+    This is not supposed to ever throw as long as the node is properly is a valid
+    data structure, even if it does not match the pattern.
+
+    This is parametric but the parameter is not used in the class. This is because
+    the parameter is meant only to statically type check that patterns are composed
+    in a way that makes sense (like not try to match a table name as a Column since
+    it is impossible). It represent the type effectively matched by the expression
+    thus covariant, not the type of the object the Pattern can try to match so that
+    Pattern[Column] is a subtype of Pattern[Expression].
+    """
+
+    @abstractmethod
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        """
+        Returns a MatchResult if the node provided matches this pattern
+        otherwise it returns None.
+
+        The parameter is not TMatchedType in that TMatchedType is the type of the
+        Node effectively matched. That would be a subtype of the type we try to match.
+        For example a Column Pattern should be able to get an Expression and match
+        a Column. TMatchedType is Column.
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class Param(Pattern[TMatchedType]):
+    """
+    Defines a named parameter in a Pattern. When matching the overall Pattern,
+    if the Pattern nested in this object matches, the corresponding input node
+    will be given a name and returned as part of the MatchResult return value.
+    This is meant to represent what a group is in a regex.
+
+    Params can be nested, and they can wrap any valid Pattern.
+    """
+
+    name: str
+    pattern: Pattern[TMatchedType]
+
+    def match(self, node: AnyType) -> Optional[MatchResult]:
+        result = self.pattern.match(node)
+        if not result:
```

Comment:
nit (it doesn't matter in this case but I think it's good practice unless there is reason _not_ to):

```suggestion
        if result is None:
```

There are a bunch of other comparisons to `None` below, I think those should probably be handled the same but I comment on all of them.

Reply:
I thought I had addressed all of them. Or maybe this was in a different PR....